Mathlib/LinearAlgebra/TensorProduct/Finite.lean

```diff
@@ -0,0 +1,170 @@
+/-
+Copyright (c) 2024 Kevin Buzzard. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kevin Buzzard
+-/
+import Mathlib.LinearAlgebra.DirectSum.Finsupp
+
+/-!
+
+# Tensor products with finite free modules
+
+We prove various results about tensoring over `R` with `Rⁿ`, if `n` is a finite type.
+
+This file contains variants of some results in `Mathlib/LinearAlgebra/DirectSum/Finsupp.lean`
+where `n →₀ R` in that file is replaced by `n → R` in this, under the hypothesis `[Finite n]`.
+
+## Main definitions
+
+* `TensorProduct.finiteLeft R M N n` : If `n` is a finite type, `Mⁿ ⊗[R] N ≃ₗ[R] (M ⊗[R] N)ⁿ`.
+* `TensorProduct.finiteRight R M N n` : If `n` is a finite type, `M ⊗[R] Nⁿ ≃ₗ[R] (M ⊗[R] N)ⁿ`.
+* `TensorProduct.finiteScalarLeft R N n` : If `n` is a finite type, `Rⁿ ⊗[R] N ≃ₗ[R] Nⁿ`.
+* `TensorProduct.finiteScalarRight R M n` : If `n` is a finite type, `M ⊗[R] Rⁿ ≃ₗ[R] Mⁿ`.
+* `TensorProduct.finiteTensorPiLid R N n m` : if `n` is a finite type, `Rⁿ ⊗[R] Nᵐ ≃ₗ[R] Nⁿˣᵐ`.
+* `TensorProduct.piTensorFiniteRid R M m n` : if `n` is a finite type, `Mᵐ ⊗[R] Rⁿ ≃ₗ[R] Mᵐˣⁿ`.
+
+Note that in the last two results, the type `m` is arbitrary and does not have to be finite.
+-/
+
+namespace TensorProduct
+
+variable (R : Type*) [CommSemiring R] (M : Type*) [AddCommMonoid M] [Module R M]
+  (N : Type*) [AddCommMonoid N] [Module R N] (ι : Type*) [DecidableEq ι] [Finite ι]
```

Comment:
```suggestion
  (N : Type*) [AddCommMonoid N] [Module R N] (ι : Type*) [Finite ι]
```
You don't need decidability for the statements to make sense, so better `open scoped Classical in` in the first two definitions.